provision/service.go

```diff
@@ -15,146 +35,166 @@ type Service interface {
+	Error       string            `json:"error,omitempty"`
 }
 
 // New returns new provision service.
-func New(cfg Config, logger logger.Logger) Service {
+func New(cfg Config, sdk SDK.SDK, logger logger.Logger) Service {
 	return &provisionService{
-		logger:           logger,
-		sdk:              cfg.SDK,
-		mfEmail:          cfg.MFEmail,
-		mfPass:           cfg.MFPass,
-		bsContent:        cfg.BSContent,
-		x509Provision:    cfg.X509Provision,
-		bsProvision:      cfg.BSProvision,
-		autoWhiteList:    cfg.AutoWhiteList,
-		predefinedThings: cfg.PredefinedThings,
+		logger: logger,
+		conf:   cfg,
+		sdk:    sdk,
 	}
 }
 
 // Provision is provision method for adding devices to proxy.
-func (ps *provisionService) Provision(externalID, externalKey string) (res Result, err error) {
-	var newThingID, ctrlChanID, dataChanID, token string
-	defer ps.recover(&err, &newThingID, &ctrlChanID, &dataChanID, &token)
-	channels := make([]string, 0)
+func (ps *provisionService) Provision(token, externalID, externalKey string) (res Result, err error) {
+	channels := make([]SDK.Channel, 0)
+	things := make([]SDK.Thing, 0)
+	defer ps.recover(&err, &things, &channels, &token)
 
-	token, err = ps.sdk.CreateToken(ps.mfEmail, ps.mfPass)
-	if err != nil {
-		return res, err
-	}
+	if token == "" {
+		token = ps.conf.Server.MfAPIKey
+		if token == "" {
+			if ps.conf.Server.MfUser == "" || ps.conf.Server.MfPass == "" {
+				return res, ErrMissingCredentials
+			}
+			u := SDK.User{
+				Email:    ps.conf.Server.MfUser,
+				Password: ps.conf.Server.MfPass,
+			}
+			token, err = ps.sdk.CreateToken(u)
+			if err != nil {
+				return res, errors.Wrap(ErrFailedToCreateToken, err)
+			}
+		}
 
-	newThingID, err = ps.sdk.CreateThing(externalID, "", token)
-	if err != nil {
-		return res, err
 	}
-
-	// Get newly created thing (in order to get the key).
-	thingCreated, err := ps.sdk.Thing(newThingID, token)
-	if err != nil {
-		return res, provsdk.ErrGetThing
+	if len(ps.conf.Things) == 0 {
+		return res, ErrEmptyThingsList
 	}
-
-	ctrlChannel, err := ps.sdk.CreateChannel("ctrlchan", "control", token)
-	if err != nil {
-		return res, provsdk.ErrCreateCtrl
+	if len(ps.conf.Channels) == 0 {
+		return res, ErrEmptyChannelsList
 	}
-	ctrlChanID = ctrlChannel.ID
-
-	dataChannel, err := ps.sdk.CreateChannel("datachan", "data", token)
-	if err != nil {
-		return res, provsdk.ErrCreateData
+	for _, thing := range ps.conf.Things {
+		// If thing in configs contains metadata with externalid
+		// set value for it from the provision request
+		if _, ok := thing.Metadata[ExternalID]; ok {
+			thing.Metadata[ExternalID] = externalID
+		}
+		th := SDK.Thing{
+			Name:     thing.Name,
+			Metadata: thing.Metadata,
+		}
+		thID, err := ps.sdk.CreateThing(th, token)
+		if err != nil {
+			res.Error = err.Error()
+			return res, errors.Wrap(ErrFailedThingCreation, err)
+		}
+		// Get newly created thing (in order to get the key).
+		thing, err := ps.sdk.Thing(thID, token)
```

Comment:
This is all confusing - you have `th` and `thing` variables because you have provision-local SDK. As noted several times - this one must be moved to global SDK.